pceplib/pcep_msg_messages_encoding.c

```diff
@@ -0,0 +1,304 @@
+        return false;
+    }
+
+    const enum pcep_object_classes *object_classes = MANDATORY_MESSAGE_OBJECT_CLASSES[msg->msg_header->type];
+    double_linked_list_node *node;
+    int index;
+    for (node = (msg->obj_list == NULL ? NULL : msg->obj_list->head), index = 0;
+         index < NUM_CHECKED_OBJECTS;
+         index++, (node = (node==NULL ? NULL : node->next_node)))
+    {
+        struct pcep_object_header *obj = ((node == NULL) ? NULL : (struct pcep_object_header*) node->data);
+
+	if ((int)object_classes[index] == NO_OBJECT) {
+		if (node != NULL) {
+			pcep_log(
+				LOG_INFO,
+				"Rejecting received message: Unexpected object [%d] present",
+				obj->object_class);
+			return false;
+		}
+	} else if (object_classes[index] != ANY_OBJECT) {
+		if (node == NULL) {
+			pcep_log(
+				LOG_INFO,
+				"Rejecting received message: Expecting object in position [%d], but none received",
+				index);
+			return false;
+		} else if (object_classes[index] != obj->object_class) {
+			pcep_log(
+				LOG_INFO,
+				"Rejecting received message: Unexpected Object Class received [%d]",
+				object_classes[index]);
+			return false;
+		}
+	}
+    }
+
+    return true;
+}
+
+struct pcep_message *pcep_decode_message(const uint8_t *msg_buf)
+{
+    uint8_t msg_version;
+    uint8_t msg_flags;
+    uint8_t msg_type;
+    uint16_t msg_length;
+
+    pcep_decode_msg_header(msg_buf, &msg_version, &msg_flags, &msg_type, &msg_length);
+
+    struct pcep_message *msg = pceplib_malloc(PCEPLIB_MESSAGES, sizeof(struct pcep_message));
+    memset(msg, 0, sizeof(struct pcep_message));
+
+    msg->msg_header = pceplib_malloc(PCEPLIB_MESSAGES, sizeof(struct pcep_message_header));
+    msg->msg_header->pcep_version = msg_version;
+    msg->msg_header->type = msg_type;
+
+    msg->obj_list = dll_initialize();
+    msg->encoded_message = pceplib_malloc(PCEPLIB_MESSAGES, msg_length);
+    memcpy(msg->encoded_message, msg_buf, msg_length);
+    msg->encoded_message_length = msg_length;
+
+    uint16_t bytes_read = MESSAGE_HEADER_LENGTH;
+    while ((msg_length - bytes_read) >= OBJECT_HEADER_LENGTH)
+    {
+        struct pcep_object_header *obj_hdr = pcep_decode_object(msg_buf + bytes_read);
+
+        if (obj_hdr == NULL)
+        {
+            pcep_log(LOG_INFO, "Discarding invalid message");
+            pcep_msg_free_message(msg);
+
+            return NULL;
+        }
+
+        dll_append(msg->obj_list, obj_hdr);
+        bytes_read += obj_hdr->encoded_object_length;
+    }
+
+    if (validate_message_objects(msg) == false)
+    {
+        pcep_log(LOG_INFO, "Discarding invalid message");
+        pcep_msg_free_message(msg);
+
+        return NULL;
+    }
+
+    return msg;
+}
+
+struct pcep_versioning *create_default_pcep_versioning()
+{
+    struct pcep_versioning *versioning = pceplib_malloc(PCEPLIB_INFRA, sizeof(struct pcep_versioning));
+    memset(versioning, 0, sizeof(struct pcep_versioning));
+
+    versioning->draft_ietf_pce_segment_routing_07 = false;
```

Comment:
this is already set to 0 on the previous line

Reply:
done